refactor the rehash implementation

slingamn opened the pull request:
This is a fairly messy refactor. The main goals are:

1. Rehashing should reload as much of the config as possible. This change adds the ability to change the MOTD, the REST API (whether it's enabled and what its listen address is), the websocket listener, and probably some other things I've forgotten.
2. Deduplicate code between `NewServer()` and `Rehash()`. With this change, `NewServer()` does not read in any config settings on its own, but rather defers to a method `applyConfig()` that it shares with `Rehash()`.
3. Get an start on fixing some of the read-write races associated with `Rehash()` by introducing `Server.configurableStateMutex`, a new `sync.RWMutex` intended to protect server state that is modified only during rehash, but read routinely by clients (e.g., the MOTD).

I have tested this change, but not exhaustively --- in particular, I didn't test that rehashing is non-disruptive to persistent websocket connections (although the documentation is pretty clear that they shouldn't be affected by `http.Server.Shutdown()`), because I don't have a websocket-capable client. Feel free to kick this back to me for more testing, or more documentation, or whatever :-)

DanielOaks commented on irc/server.go:
```diff
@@ -1453,6 +1318,18 @@ func (server *Server) rehash() error {
+
+	server.name = config.Server.Name
+	server.nameCasefolded = casefoldedName
+	server.networkName = config.Network.Name
```
We should make sure that server name can't change after launching the server, and error out with something like the maximum linelen message above. I'd be okay with the network name changing, so long as a new `NETWORK=blah` RPL_ISUPPORT token is sent (should be but good to confirm).

DanielOaks commented:
Nice, this looks really cool! Aside from that server name comment I've added, from a glance this looks good.

Honestly, I've been thinking about completely removing the websocket server stuff, at least until we clone how one of the other major servers does it (Insp, Unreal and Chary all have websocket implementations, see also https://github.com/ircv3/ircv3-specifications/issues/310 ). Since I don't think anything uses it and we don't need to keep it around, thoughts on ripping it out?

slingamn commented:
Yeah, removal sounds good to me. I'll include that in my next commit.

DanielOaks commented:
Awesome, thanks very much!

slingamn commented:
Not cleaned up yet: references to `github.com/gorilla/websocket` in `Gopkg.lock`, `Gopkg.toml`, and `vendor`.

slingamn commented:
I tried changing the network name and got this output line:

````
2017-09-28T06:58:56Z : debug : useroutput : shivaram_hexchat :  -> : @time=2017-09-28T06:58:56.134Z :oragono.test 005 shivaram_hexchat NETWORK=OragonoTest.changed :are supported by this server
````
Hexchat seems to have interpreted it correctly (the sidebar name changed from `OragonoTest` to `OragonoTest.changed`).

DanielOaks commented:
Awesome, in that case I think that's all the nits I have for this! I'll merge this in pretty soon and throw my own stuff on top of it, thanks very much for cleaning this up.
